### PR TITLE
Implement vec_coerce() generic and methods

### PR DESCRIPTION
**Pull request by lionel-**

### CODE REVIEW

File: R/vec-coerce.R [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    factor(from, levels(fct))
+  }
+)
+
+def_method2("factor", "factor",
+  vec_coerce = function(from, to, ...) {
+    lvls_from <- levels(from)
+    lvls_to <- levels(to)
+
+    if (!(lvls_from %in% lvls_to) && !(lvls_to %in% lvls_from)) {
```
Comment (hadley) [inline R/vec-coerce.R]: `setequal()`?

Comment (hadley) [reply R/vec-coerce.R]: Actually, I don't understand what you're doing here - where does the vector become a scalar?

File: R/vec-coerce.R [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      msg <- "Coercing `factor` to `character` because of incompatible levels"
+      warn(msg, "rlang_vec_coerce_wng")
+      return(as.character(from))
+    }
+
+    n_from <- length(lvls_from)
+    n_to <- length(lvls_to)
+    if (n_from != n_to) {
```
Comment (hadley) [inline R/vec-coerce.R]: How can levels be congruent but not the same length?

File: R/vec-coerce.R [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+)
+
+def_method2("factor", "factor",
+  vec_coerce = function(from, to, ...) {
+    lvls_from <- levels(from)
+    lvls_to <- levels(to)
+
+    if (!(lvls_from %in% lvls_to) && !(lvls_to %in% lvls_from)) {
+      msg <- "Coercing `factor` to `character` because of incompatible levels"
+      warn(msg, "rlang_vec_coerce_wng")
```
Comment (hadley) [inline R/vec-coerce.R]: Don't we want to warn when generating the common type, rather than doing the coercion?

File: R/vec-coerce.R [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+## Export all methods
+#' @rawNamespace export(.__rlang_binary_strict_methods__.)
+NULL
+
+
+### Numeric coercions
+
+def_method2("logical", "logical",
```
Comment (hadley) [inline R/vec-coerce.R]: Maybe we could define self-to-self with `walk()` to make it that pass more clear?

File: R/vec-coerce.R [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    from
+  }
+)
+def_method2("integer", "numeric",
+  vec_coerce = function(from, to, ...) {
+    vec_coerce_bare(from, "numeric")
+  }
+)
+
+def_method2("numeric", "numeric",
```
Comment (hadley) [inline R/vec-coerce.R]: We also need numeric -> integer etc?

Comment (lionel-) [reply R/vec-coerce.R]: They are already defined as method definition is symmetric.

File: R/vec-coerce.R (hunk 4 above)
Comment (lionel-) [reply R/vec-coerce.R]: what do you mean?

File: R/vec-coerce.R (hunk 3 above)
Comment (lionel-) [reply R/vec-coerce.R]: For simplicity I thought we'd generate the common type with coercion warnings muffled, then we can use the same generic / methods.

File: R/vec-coerce.R (hunk 2 above)
Comment (lionel-) [reply R/vec-coerce.R]: when one is a subset of the other. I think it's helpful to provide a more precise warning in that case?

File: R/vec-coerce.R (hunk 4 above)
Comment (hadley) [reply R/vec-coerce.R]: Like 

```R
types <- c("integer", "character", "double")

for (type in types) {
  def_method2(type, type, function(x, y) x)
}
```

File: R/vec-coerce.R (hunk 5 above)
Comment (hadley) [reply R/vec-coerce.R]: Oh hmmmm, now I wonder if that's going to be confusing

File: R/vec-coerce.R (hunk 2 above)
Comment (hadley) [reply R/vec-coerce.R]: That sounds correct; I just find the current logic in your code hard to follow.

File: R/vec-coerce.R (hunk 3 above)
Comment (hadley) [reply R/vec-coerce.R]: I think the long run, I don't think that will be tenable because the performance will be too slow.

File: R/vec-coerce.R (hunk 5 above)
Comment (lionel-) [reply R/vec-coerce.R]: You're going to get a warning if you try to redefine the method, which helps the implementer but maybe not the reader. We should discuss whether we'd like binary dispatch to be more general and allow non-symmetry, then maybe we need an option or different function name reflecting the symmetric case.

File: R/vec-coerce.R (hunk 2 above)
Comment (lionel-) [reply R/vec-coerce.R]: sorry, how would you do it?

Comment (hadley) [reply R/vec-coerce.R]: I'd write it more like this:

```R
if (setequal(lvls_from, lvls_to)) {
  factor(from, lvls_to)
} else if (all(lvls_from %in% lvls_to)) {
  warn(...)
  factor(from, lvls_to)
} else {
  warn(...)
  as.character(from)
}
```

Comment (lionel-) [reply R/vec-coerce.R]: Thanks it's much better. I'd just always use the longer levels rather than `lvls_to` in the subset case so that coercion behaviour is symmetric (and it's the choice that preserves the most information).

Comment (lionel-) [reply R/vec-coerce.R]: Well actually that's not right: `all(lvls_from %in% lvls_to)` is FALSE when LHS is longer than rhs. How about:

```r
    union <- union(lvls_from, lvls_to)
    lengths <- c(length(lvls_from), length(lvls_to))
    congruent <- length(union) == lengths

    if (all(congruent)) {
      from
    } else if (any(congruent)) {
      msg <- "Factor levels are congruent but not the same length"
      warn(msg, "rlang_vec_coerce_wng")
      lvls <- if (lengths[[1]] > lengths[[2]]) lvls_from else lvls_to
      factor(from, lvls)
    } else {
      msg <- "Coercing `factor` to `character` because of incompatible levels"
      warn(msg, "rlang_vec_coerce_wng")
      as.character(from)
    }
```

Comment (hadley) [reply R/vec-coerce.R]: I don't think that logic is correct either - it might be better to just create `lvls_min` and `lvls_max`
